src/utils/hash.ts

```diff
@@ -3,60 +3,65 @@ import { get, post } from "./http";
 const _sodium = require('libsodium-wrappers');
 let sodium: any;
 
-let userHashedValues: any = {};
-
-export async function hashValue(value: string, dataType: string, jwt?: string) {
-  if (!jwt) {
-    return ""
-  }
-
-  if (!value) {
-    return value;
-  }
+export let userHashedValues: any = {};
 
+export async function hashValues(payload: any, jwt: string) {
   if (sodium === undefined) {
     await Promise.all([setUserHashedValues(jwt), _sodium.ready]).then(() => {
       sodium = _sodium;
     });
   }
 
-  const hashedValue = sodium.to_hex(sodium.crypto_generichash(64, value));
+  let result: any = {}
 
-  await encryptValue(value, hashedValue, dataType, jwt);
+  await payload.forEach(async (hashPayload: any) => {
+    if(hashPayload.value?.length) {
+      const hashedValue = sodium.to_hex(sodium.crypto_generichash(64, hashPayload.value));
+      const alreadyEncrypted = await hasHashValueInCache(hashPayload.dataType, hashedValue);
+
+      if(!alreadyEncrypted) {
+        await encryptValue(hashPayload.value, hashedValue, hashPayload.dataType, jwt)
+      }
+
+      result[hashPayload.dataType] = hashedValue;
+    } else {
+      result[hashPayload.dataType] = hashPayload.value;
+    }
+  });
 
-  return hashedValue;
+  return result;
 }
 
-async function encryptValue(value: string, hashedValue: string, dataType: string, jwt?: string) {
-  if (!jwt) {
-    return ""
+export async function hasHashValueInCache(dataType: string, hashedValue: string) {
+  // add dataType to userHashedValues cache if it doesn't already exist
+  if(!userHashedValues[dataType]) {
+    userHashedValues[dataType] = []
   }
 
-  const hashValueAlreadyExists = !!userHashedValues[dataType] && userHashedValues[dataType].includes(hashedValue);
+  const hasValue = userHashedValues[dataType].includes(hashedValue);
 
-  if(!hashValueAlreadyExists) {
-    const params = {
-      value: value,
-      hashed_value: hashedValue,
-      data_type: dataType
-    }
-
-    const response =  await post("/user_encrypted_data", params, jwt);
+  // if it does not have the value, add it
+  if(!hasValue) {
+    userHashedValues[dataType].push(hashedValue)
+  }
 
-    if(response.status == 201) {
-      // update the local userHashedValues so that we don't re-encrypt it
-      if(userHashedValues[dataType]) {
-        userHashedValues[dataType].push(hashedValue);
-      } else {
-        userHashedValues[dataType] = [hashedValue]
-      }
-    }
+  // return the result of hasValue
+  return hasValue
+}
 
-    setUserHashedValues(jwt);
+async function encryptValue(value: string, hashedValue: string, dataType: string, jwt: string) {
+  const params = {
+    value: value,
+    hashed_value: hashedValue,
+    data_type: dataType
   }
+
+  post("/user_encrypted_data", params, jwt).then(async (_resp) => {
+    await setUserHashedValues(jwt)
```

Comment:
you can just await the post call. the await in the thenable will not hold off any subsequent calls